service/src/main/java/org/broadinstitute/listener/relay/http/RelayedHttpRequestProcessor.java

```diff
@@ -83,6 +97,38 @@ public Result writeNotAcceptedResponseOnCaller(RelayedHttpListenerContext contex
     return Result.FAILURE;
   }
 
+  public Result writePreflightResponse(RelayedHttpListenerContext context) {
+    if (context.getResponse() == null) {
+      logger.error("The context did not have a valid response");
+      return Result.FAILURE;
+    }
+
+    RelayedHttpListenerResponse listenerResponse = context.getResponse();
+    listenerResponse.setStatusCode(204);
+    listenerResponse
+        .getHeaders()
+        .put(ACCESS_CONTROL_ALLOW_METHODS, corsSupportProperties.preflightMethods());
+
+    if (!listenerResponse.getHeaders().containsKey(ACCESS_CONTROL_ALLOW_ORIGIN))
+      listenerResponse
+          .getHeaders()
+          .put(
+              ACCESS_CONTROL_ALLOW_ORIGIN,
+              context.getRequest().getHeaders().getOrDefault("Origin", "*"));
+
+    listenerResponse.getHeaders().put(ACCESS_CONTROL_ALLOW_CREDENTIALS, "true");
+    listenerResponse
+        .getHeaders()
+        .put(ACCESS_CONTROL_ALLOW_HEADERS, "Authorization, Content-Type, Accept, Origin,X-App-Id");
+    listenerResponse.getHeaders().put(ACCESS_CONTROL_MAX_AGE, "1728000");
```

Comment:
We should make the allowed headers and age configuration properties - in another ticket.